controlplane/kubeadm/internal/machinefilters/machine_filters.go

```diff
@@ -277,8 +277,14 @@ func matchClusterConfiguration(kcp *controlplanev1.KubeadmControlPlane, machine
 	if ok {
 		machineClusterConfig := &kubeadmv1.ClusterConfiguration{}
 		// ClusterConfiguration annotation is not correct, only solution is to rollout.
-		if err := json.Unmarshal([]byte(machineClusterConfigStr), machineClusterConfig); err != nil {
-			return false
+		// NOTE: it is required to handle a nil ClusterConfiguration (that is serialized into "null") as a special case,
+		// because Unmarshal returns &kubeadmv1.ClusterConfiguration{}, not nil; see https://github.com/kubernetes-sigs/cluster-api/issues/3353
+		if machineClusterConfigStr != "null" {
+			if err := json.Unmarshal([]byte(machineClusterConfigStr), machineClusterConfig); err != nil {
+				return false
+			}
```

Comment:
```suggestion
			if err := json.Unmarshal([]byte(machineClusterConfigStr), &machineClusterConfig); err != nil {
				return false
			}
```

This should be enough to make `machineClusterConfig` be set to nil.

Reply:
I was thinking the same thing, this demonstrates the slight problem with nil types: https://play.golang.org/p/RHhKGTgyGbU

Reply:
done!